### ISSUE TITLE
Allow adding a "qualification line" to the resolver

### ISSUE THREAD

**Issue by niemela** [icpctools/icpctools#983]
At many events one of the most important aspect of the results is who qualifies to the next level of contest. It would be nice to be able to visualize this.

In some instances this number can be quite large. E.g. at the NAC it would often be ~16. For this reason a good way to show it would probably be some kind if line drawn between teams at rank `n` and `n+1` where `n` is the number.

**Comment by deboer-tim** [comment]
Many years ago at finals we would show this for the medals by having rows with gold/silver/bronze backgrounds, kind of like how the resolver judge queue has a set of green rows at the top. Think this would work too? I think it might be a little more visual.

IMHO in either case I think it would be best if we could keep this associated with an 'award', so that it is in the feed, avoids another command line argument, and comes with a citation like 'advances to finals'.

We do have a challenge that like `b` / # of bronzes, contest directors often have a little flexibility to change the number of teams, or will base it on how close the team 'below the line' does, and they often like to do that live on stage - i.e. if the contest is supposed to have 5 teams advancing they wouldn't want a line drawn at 6 teams and give away the fact that we're going to announce team 6 during the resolution. Maybe there is something we can do here with the ordering of the award vs contest state that would mean it should also be part of the resolution?

**Comment by niemela** [comment]
> Many years ago at finals we would show this for the medals by having rows with gold/silver/bronze backgrounds, kind of like how the resolver judge queue has a set of green rows at the top. Think this would work too? I think it might be a little more visual.

I agree that that would be more visual. My worry is that it would be too much.

> IMHO in either case I think it would be best if we could keep this associated with an 'award', so that it is in the feed, avoids another command line argument, 

I'm a little bit torn about this, but I suspect you are right.

The counterargument is that having command line options is arguably easier for the lightweight users. I.e. the ones that don't have a CDS or any other part of the infrastructure, and only wants to run the resolver.

> and comes with a citation like 'advances to finals'.

Does "comes with a citation" imply that there must be a "stop" at that team? Because we definitely don't want that.

> We do have a challenge that like `b` / # of bronzes, contest directors often have a little flexibility to change the number of teams, or will base it on how close the team 'below the line' does, and they often like to do that live on stage - i.e. if the contest is supposed to have 5 teams advancing they wouldn't want a line drawn at 6 teams and give away the fact that we're going to announce team 6 during the resolution. Maybe there is something we can do here with the ordering of the award vs contest state that would mean it should also be part of the resolution?

Even when that happens (like e.g. last year at the NAC) I would argue that it is better to show the nominal number of teams in the resolver and just announce the change.

That said.... it would certainly be *very* cool if the line is at position N, and the *after* passing it we stop and then it extends (i.e. matching the announcement). It's starting to get really complicated now though...

**Comment by deboer-tim** [comment]
> > Many years ago at finals we would show this for the medals by having rows with gold/silver/bronze backgrounds, kind of like how the resolver judge queue has a set of green rows at the top. Think this would work too? I think it might be a little more visual.
> 
> I agree that that would be more visual. My worry is that it would be too much.

Ok.

> > IMHO in either case I think it would be best if we could keep this associated with an 'award', so that it is in the feed, avoids another command line argument,
> 
> I'm a little bit torn about this, but I suspect you are right.
> 
> The counterargument is that having command line options is arguably easier for the lightweight users. I.e. the ones that don't have a CDS or any other part of the infrastructure, and only wants to run the resolver.

My hope for the award utility (command line or UI) was that you could run it once to assign awards, ie same usability but at least the awards are stored in a file, more likely to be backed up, etc. and same flow works for more complex cases.

> > and comes with a citation like 'advances to finals'.
> 
> Does "comes with a citation" imply that there must be a "stop" at that team? Because we definitely don't want that.

No, it doesn't. You can already pause, go to award screen, show list, or ignore for any award.

> > We do have a challenge that like `b` / # of bronzes, contest directors often have a little flexibility to change the number of teams, or will base it on how close the team 'below the line' does, and they often like to do that live on stage - i.e. if the contest is supposed to have 5 teams advancing they wouldn't want a line drawn at 6 teams and give away the fact that we're going to announce team 6 during the resolution. Maybe there is something we can do here with the ordering of the award vs contest state that would mean it should also be part of the resolution?
> 
> Even when that happens (like e.g. last year at the NAC) I would argue that it is better to show the nominal number of teams in the resolver and just announce the change.
> 
> That said.... it would certainly be _very_ cool if the line is at position N, and the _after_ passing it we stop and then it extends (i.e. matching the announcement).

Agreed. A bit odd if we say you advance to finals/got a medal and didn't show it on screen, very cool if it happened dynamically.

> It's starting to get really complicated now though...

Is this before or after resolving multiple contests with new awards and screens and animations on multiple clients in the desert? ;-) It seems like every year we come up with another thing that has changed or would be useful if we could add.

**Comment by niemela** [comment]
> My hope for the award utility (command line or UI) was that you could run it once to assign awards, ie same usability but at least the awards are stored in a file, more likely to be backed up, etc. and same flow works for more complex cases.

Yes, and it's awesome. 👍 

One (minor) issue is that you can't run the awards utility when reading the feed from an API. (Of course, the trivial workaround is to download the feed from the API). 

> > Does "comes with a citation" imply that there must be a "stop" at that team? Because we definitely don't want that.
> 
> No, it doesn't. You can already pause, go to award screen, show list, or ignore for any award.

👍 

> > It's starting to get really complicated now though...
> 
> Is this before or after resolving multiple contests with new awards and screens and animations on multiple clients in the desert? ;-) 

Yes. 😃
